Microsoft.Identity.Web/Client/TokenAcquisition.cs

```diff
@@ -125,9 +125,18 @@ public async Task AddAccountToCacheFromAuthorizationCode(AuthorizationCodeReceiv
             try
             {
                 // As AcquireTokenByAuthorizationCodeAsync is asynchronous we want to tell ASP.NET core that we are handing the code
-                // even if it's not done yet, so that it does not concurrently call the Token endpoint.
+                // even if it's not done yet, so that it does not concurrently call the Token endpoint. (otherwise there will be a
+                // race condition ending-up in an error from Azure AD telling "code already redeemed")
                 context.HandleCodeRedemption();
 
+                // The cache will need the claims from the ID token. In the case of guest scenarios
```

Comment:
That's the core of the proposed fix. Make sure that the Http.Context.User has the claims.